jib-core/src/main/java/com/google/cloud/tools/jib/api/RegistryImage.java

```diff
@@ -0,0 +1,125 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.api;
+// TODO: Move to com.google.cloud.tools.jib once that package is cleaned up.
+
+import com.google.cloud.tools.jib.configuration.credentials.Credential;
+import com.google.cloud.tools.jib.configuration.credentials.CredentialRetriever;
+import com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory;
+import com.google.cloud.tools.jib.image.ImageReference;
+import com.google.cloud.tools.jib.image.InvalidImageReferenceException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Builds to a container registry.
+ *
+ * <p>The registry portion of the image reference determines which registry to push the image to.
+ * The repository portion is the namespace to push the image to. The tag is a label to easily
+ * identify an image among all the images in the repository. See {@link ImageReference} for more
+ * details.
+ *
+ * <p>When configuring credentials (via {@link #setCredential} for example), make sure the
+ * credentials are valid push credentials for the repository specified via the image reference.
+ */
+// TODO: Add tests once JibContainerBuilder#containerize() is added.
+public class RegistryImage {
+
+  /**
+   * Instantiate with the image reference to push to.
+   *
+   * @param imageReference the image reference
+   * @return a new {@link RegistryImage}
+   */
+  public static RegistryImage named(ImageReference imageReference) {
+    return new RegistryImage(imageReference);
+  }
+
+  /**
+   * Instantiate with the image reference to push to.
+   *
+   * @param imageReference the image reference
+   * @return a new {@link RegistryImage}
+   * @throws InvalidImageReferenceException if {@code imageReference} is not a valid image reference
+   */
+  public static RegistryImage named(String imageReference) throws InvalidImageReferenceException {
+    return named(ImageReference.parse(imageReference));
+  }
+
+  private final ImageReference imageReference;
+  private List<CredentialRetriever> credentialRetrievers = new ArrayList<>();
+
+  /** Instantiate with {@link #named}. */
+  private RegistryImage(ImageReference imageReference) {
+    this.imageReference = imageReference;
+  }
+
+  /**
+   * Sets the credentials to use to push the image. This replaces any previously-added {@link
+   * CredentialRetriever}s.
+   *
+   * @param username the username
+   * @param password the password
+   * @return this
+   */
+  public RegistryImage setCredential(String username, String password) {
+    credentialRetrievers = new ArrayList<>();
+    credentialRetrievers.add(() -> Optional.of(Credential.basic(username, password)));
```

Comment:
Just wondering why Credential.basic would ever return null?

Reply:
`Credential.basic` would not return null, but `CredentialRetriever` can return an optional `Credential`.